### ISSUE TITLE
warning  in ./node_modules/peerjs/dist/peerjs.min.js  Critical dependency: the request of a dependency is an expression

### ISSUE THREAD

**Issue by marcoippolito** [peers/peerjs#630]
My testing-app is compiling fine, except that I get this warning:
" Critical dependency: the request of a dependency is an expression"

    (base) marco@pc01:~/webMatters/vueMatters/PeerJS-VueJS-Test$ npm run serve

    > testproject@0.1.0 serve /home/marco/webMatters/vueMatters/PeerJS-VueJS-Test
    > vue-cli-service serve

     INFO  Starting development server...
    98% after emitting CopyPlugin

    WARNING  Compiled with 1 warnings                                                                                                             
    7:22:25 PM

    warning  in ./node_modules/peerjs/dist/peerjs.min.js

    Critical dependency: the request of a dependency is an expression

      App running at:
      - Local:   http://localhost:8080 
      - Network: http://ggc.world/

      Note that the development build is not optimized.
      To create a production build, run npm run build.

This is webpack.config.js  :  

    {
        "mode": "development",
        "output": {
            "path": __dirname+'/static',
            "filename": "[name].[chunkhash:8].js"
        },
        "module": {
            "rules": [
                {
                    "test": /\.vue$/,
                    "exclude": /node_modules/,
                    "use": "vue-loader"
                },
                {
                    "test": /\.pem$/,
                    "use": "file-loader"
                }
            ]
        },
        node: {
            __dirname: false,
            __filename: false
        },
        resolve: {
            extension: ['*', '.pem'],
        },
        devServer: {
            watchOptions: {
                aggregateTimeout: 300,
                poll: 1000
            },
            https: true,
            compress: true,
            public: 'ggc.world:8080'
        }
    }

Any ideas about how to solve it?

**Comment by afrokick** [comment]
Similar to https://github.com/peers/peerjs/issues/537

**Comment by jextrevor** [comment]
I'm also having this issue when using create-react-app with peerjs.

**Comment by jgresham** [comment]
I am having this issue as well with both peerjs v0.3.18 and latest version.

"react": "^16.13.1"
"react-scripts": "3.4.1"

**Comment by loclv** [comment]
I'm also having this issue when using:

```json
{
    "peerjs": "^1.3.1",
    "webpack": "^4.43.0",
    "webpack-cli": "^3.3.12"
}
```

**Comment by jextrevor** [comment]
> I'm also having this issue when using:
> 
> ```json
> {
>     "peerjs": "^1.3.1",
>     "webpack": "^4.43.0",
>     "webpack-cli": "^3.3.12"
> }
> ```

I might be wrong, but I believe it's recommended to install webpack-cli globally and not as part of a project

**Comment by loclv** [comment]
@jextrevor 
> I might be wrong, but I believe it's recommended to install webpack-cli globally and not as part of a project

Thank you!

**Comment by mbsyaswanth** [comment]
> I'm also having this issue when using create-react-app with peerjs.

I am getting the same warning in create-react-app .. Any updates?

for some reason, I am getting `peer.call(userId, selfStream);` as undefined.
please help me - the application is [here](https://video-meet.netlify.app/). Create a room and try to join a person by copy pasting the meeting url in the address bar. Then you will see the error in console

**Comment by joshverd** [comment]
Still getting this issue over here.

**Comment by Pigotz** [comment]
I have this issue too :pensive:

**Comment by ElonLotfi** [comment]
I have this issue

**Comment by ntngoc96** [comment]
Me toooo.

**Comment by preciousbetine** [comment]
same here

**Comment by preciousbetine** [comment]
What I ended up doing was to remove the peerjs import from App.js and Include a script tag in my index.html file. For eslint, I added /* eslint-disable no-undef */ at the top of the file.

**Comment by canotur** [comment]
I have this issue too. Peerjs is not working with Angular? Does not PeerJs support Angular, or wont? 
Is there any working Angular Project which using Peerjs?

**Comment by deepakpixel** [comment]
Is it still not fixed? This warning is so annoying.

WARNING:
./node_modules/peerjs/dist/peerjs.min.js
Critical dependency: the request of a dependency is an expression

I am using "peerjs": "^1.3.2" with ReactJS (create-react-app)

**Comment by Pigotz** [comment]
For anyone landing on this issue, I came up with a possible (not so clean) solution last night: have you ever heard of [`patch-package`](https://www.npmjs.com/package/patch-package)?

**Keep in mind that the solution shown below works with `peerjs@1.3.2` but I do not guarantee that the same patch is going to work with future versions of the package**

## TL;DR

1. Install [`patch-package`](https://www.npmjs.com/package/patch-package) inside `devDependencies`
2. Add a `postinstall` script inside your project's `package.json` with the following: `patch-package`
3. Download and unzip this patch ([peerjs+1.3.2.patch.zip](https://github.com/peers/peerjs/files/7089083/peerjs%2B1.3.2.patch.zip)) and place it inside a `patches` folder inside the root of your project
4. Remove all the `node_modules`
5. Perform a clean `npm i` or `yarn`, depending on the package manager you use
6. Enjoy

## Why

First of all: because that warning during the compilation of my project with Webpack is quite annoying.

Secondly: there is no real need for PeerJS to be compiled with Parcel, unless you're planning to build your project with Parcel as well but the majority of projects don't.

Also, I'm wondering why the authors of this library are not going to provide support for multiple compiled versions of the library.
I mean: it shouldn't be painful to support a normal TS compilation of the package **and** provide a bundled version as well, like they already do with Parcel.

This way, if a developer needs to use this package inside a bundled project, they would be free to import it as a normal ES5 library. They would still be free to import the minified bundled version as well.

## How

_I'm going to provide you the steps behind that patch, in order to let anyone better understand how it's been made._

- I `cd`ed inside the `node_modules/peerjs`
- I created a base `tsconfig.json` file with `npx tsc --init`
- I changed some flags inside the default `tsconfig.json` configuration, _not necessary though_
- I placed a lot of `// @ts-ignore` comments inside the `.ts` files in the `lib` folder, because I didn't want to change the actual source code but the compiler was complaining about some issues
- I had to create an `index.ts` file inside the `lib` folder which `export * from './all_the_other_files'`, _check it inside the patch_
- I also had to create a `peerjs.min.ts` inside the `lib` folder which exports everything from the `index` one, the reason is that `patch-package` doesn't allow me to patch the package.json _(in order to change the `main` property to something else)_ for some security reason, relatable
- I removed al the files inside the `dist` folder
- I prepared a `build:ts` script inside the `package.json` that simply runs `npx tsc`
- I built the package with the previous script
- I created the patch file by running `npx patch-package peerjs` inside the root of my project

**Comment by siddharthmagadum16** [comment]
Hi, @Pigotz 
I was trying your method to fix this. But after following those 5 steps, I face this error in browser, when I do `npm start`
`Uncaught TypeError: peerjs__WEBPACK_IMPORTED_MODULE_2___default.a is not a constructor`

I'm using react v17, peerjs v1.3.2, patch-package v6.4.7

**Comment by Pigotz** [comment]
@siddharthmagadum16 you're right, try this other patch: [peerjs+1.3.2.patch.zip](https://github.com/peers/peerjs/files/7127004/peerjs%2B1.3.2.patch.zip)

**Comment by siddharthmagadum16** [comment]
@Pigotz  Yes, it works, thank you!

**Comment by Hashen110** [comment]
hey @Pigotz 

I got an error after removing node_modules and `npm install`

> core-js@3.18.3 postinstall C:\Users\HP\Projects\twiists\twiists_frontend\node_modules\core-js
> node -e "try{require('./postinstall')}catch(e){}"

Thank you for using core-js ( https://github.com/zloirock/core-js ) for polyfilling JavaScript standard library!

The project needs your help! Please consider supporting of core-js:
> https://opencollective.com/core-js 
> https://patreon.com/zloirock 
> https://paypal.me/zloirock 
> bitcoin: bc1qlea7544qtsmj2rayg0lthvza9fau63ux0fstcz 

Also, the author of core-js ( https://github.com/zloirock ) is looking for a good job -)


> ejs@2.7.4 postinstall C:\Users\HP\Projects\twiists\twiists_frontend\node_modules\ejs
> node ./postinstall.js

Thank you for installing EJS: built with the Jake JavaScript build tool (https://jakejs.com/)


> twiists_frontend@0.0.1 postinstall C:\Users\HP\Projects\twiists\twiists_frontend
> patch-package peerjs

patch-package 6.4.7
• Creating temporary folder
• Installing peerjs@1.3.2 with npm
• Diffing your files with clean files
⁉️  Not creating patch file for package 'peerjs'
⁉️  There don't appear to be any changes.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! twiists_frontend@0.0.1 postinstall: `patch-package peerjs`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the twiists_frontend@0.0.1 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\HP\AppData\Roaming\npm-cache\_logs\2021-10-17T15_16_50_122Z-debug.log


package.json:
{
  "scripts": {
    "lint": "eslint --ext .js,.vue ./",
    "test": "echo \"No test specified\" && exit 0",
    "postinstall": "patch-package peerjs"
  },
  "dependencies": {
    "@quasar/extras": "^1.0.0",
    "axios": "^0.21.1",
    "core-js": "^3.6.5",
    "peerjs": "^1.3.2",
    "quasar": "^2.0.0",
    "socket.io-client": "^4.3.2",
    "socket.io-msgpack-parser": "^3.0.1"
  },
  "devDependencies": {
    "@babel/eslint-parser": "^7.13.14",
    "@quasar/app": "^3.0.0",
    "eslint": "^7.14.0",
    "eslint-config-standard": "^16.0.2",
    "eslint-plugin-import": "^2.19.1",
    "eslint-plugin-node": "^11.0.0",
    "eslint-plugin-promise": "^5.1.0",
    "eslint-plugin-vue": "^7.0.0",
    "eslint-webpack-plugin": "^2.4.0",
    "patch-package": "^6.4.7"
  },
  "browserslist": [
    "last 30 Chrome versions",
    "last 30 Firefox versions",
    "last 12 Edge versions",
    "last 21 Safari versions",
    "last 24 Android versions",
    "last 24 ChromeAndroid versions",
    "last 24 FirefoxAndroid versions",
    "last 30 iOS versions",
    "last 15 Opera versions"
  ],
  "engines": {
    "node": ">= 12.22.1",
    "npm": ">= 6.13.4",
    "yarn": ">= 1.21.1"
  }
}

**Comment by dyingsunlight** [comment]
If you are using peerjs with typescript , import from source directly may resolve the issue

```typescript
import { Peer } from 'peerjs/lib/peer'

const local = new Peer()
```

This work fine for me

**Comment by apare** [comment]
To fix this problem I added this to my webpack config:
```js
module.exports = {
  module: {
    exprContextCritical: false
  }
}
```
This solve the warning but I'm not sure if `exprContextCritical` is important for you

**Comment by yelban** [comment]
> ```ts
> import { Peer } from 'peerjs/lib/peer'
> ```

This work fine in Javascript

**Comment by yelban** [comment]
> To fix this problem I added this to my webpack config:
> 
> ```js
> module.exports = {
>   module: {
>     exprContextCritical: false
>   }
> }
> ```
> 
> This solve the warning but I'm not sure if `exprContextCritical` is important for you

This work fine in Javascript

**Comment by artola** [comment]
> ```js
> exprContextCritical
> ```

@apare Here my use with a more restricted application:

```js
      new webpack.ContextReplacementPlugin(/peerjs/, (context) => {
        context.dependencies[0].critical = false;

        return context;
      }),
```

**Comment by afrokick** [comment]
Should be fixed in peerjs@beta

**Comment by terragady** [comment]
the peerjs@beta does not fix it. I get a new error:
```
Uncaught (in promise) TypeError: Peer is not a constructor
    at e.value (Assist.js:193:34)
    at Assist.js:49:18
    at index.js:285:49
    at Array.forEach (<anonymous>)
    at index.js:285:33
```

**Comment by adarsh-chakraborty** [comment]
Still not fixed
